File: scripts/re/ghidra_extract_fw_flow.py
#@author codex
#@category MT7927
#@keybinding
#@menupath
#@toolbar

from ghidra.program.model.listing import Function
from ghidra.program.model.symbol import RefType
from java.lang import String

# Args:
# 1) output markdown path
out_path = None
args = getScriptArgs()
if args and len(args) > 0:
    out_path = args[0]
else:
    out_path = "/tmp/mtkwecx_fw_flow.md"

keywords = [
    "MT6639PreFirmwareDownloadInit",
    "MT6639WpdmaConfig",
    "MT6639InitTxRxRing",
    "MT6639ConfigIntMask",
    "HalDownloadPatchFirmware",
    "AsicConnac3xLoadRomPatch",
    "AsicConnac3xLoadFirmware",
    "P0T15:AP CMD",
    "P0T16:FWDL",
    "P0R0:FWDL",
    "HOST_INT_STA",
    "MCU_CMD",
    "WPDMA_GLO_CFG",
    "WFDMA",
]

listing = currentProgram.getListing()
mem = currentProgram.getMemory()

matches = []
visited = set()

it = listing.getDefinedData(True)
while it.hasNext() and not monitor.isCancelled():
    d = it.next()
    val = d.getValue()
    if val is None:
        continue
    try:
        s = str(val)
    except:
        continue
    if len(s) < 4:
        continue

    hit = False
    for kw in keywords:
        if kw in s:
            hit = True
            break
    if not hit:
        continue

    saddr = d.getAddress()
    key = str(saddr)
    if key in visited:
        continue
    visited.add(key)

    refs = getReferencesTo(saddr)
    if refs is None:
        refs = []

    ref_rows = []
    for r in refs:
        from_addr = r.getFromAddress()
        fn = getFunctionContaining(from_addr)
        fn_name = fn.getName() if fn else "<no_function>"
        entry = str(fn.getEntryPoint()) if fn else ""

        ins = listing.getInstructionAt(from_addr)
        ins_text = ""
        if ins:
            ins_text = ins.toString()

        ref_rows.append((str(from_addr), fn_name, entry, ins_text, str(r.getReferenceType())))

    matches.append((str(saddr), s, ref_rows))

# sort by address string for deterministic output
matches.sort(key=lambda x: x[0])

# collect interesting function set
func_map = {}
for m in matches:
    for rr in m[2]:
        fn_name = rr[1]
        fn_entry = rr[2]
        if fn_name == "<no_function>" or fn_entry == "":
            continue
        func_map[fn_entry] = fn_name

# helper: collect callees and callers for each function
call_graph = {}
for fentry, fname in func_map.items():
    faddr = toAddr(fentry)
    fn = getFunctionAt(faddr)
    if fn is None:
        continue

    callees = set()
    callers = set()

    body = fn.getBody()
    ins_it = listing.getInstructions(body, True)
    while ins_it.hasNext() and not monitor.isCancelled():
        ins = ins_it.next()
        ft = ins.getFlowType()
        if ft and ft.isCall():
            flows = ins.getFlows()
            if flows:
                for fa in flows:
                    cfn = getFunctionContaining(fa)
                    if cfn:
                        callees.add(cfn.getName() + "@" + str(cfn.getEntryPoint()))
                    else:
                        callees.add(str(fa))

    refs_to = getReferencesTo(fn.getEntryPoint())
    if refs_to:
        for r in refs_to:
            if r.getReferenceType().isCall():
                cfrom = r.getFromAddress()
                cfn = getFunctionContaining(cfrom)
                if cfn:
                    callers.add(cfn.getName() + "@" + str(cfn.getEntryPoint()))
                else:
                    callers.add(str(cfrom))

    call_graph[fentry] = (fname, sorted(list(callers)), sorted(list(callees)))

# write markdown output
f = open(out_path, "w")
f.write("# mtkwecx.sys firmware/DMA reverse notes (auto extract)\n\n")
f.write("Program: %s\n\n" % currentProgram.getName())

f.write("## String Xrefs\n\n")
for saddr, sval, refs in matches:
    f.write("### %s `%s`\n\n" % (saddr, sval.replace("`", "'")))
    if not refs:
        f.write("- (no xrefs found)\n\n")
        continue
    for rr in refs:
        f.write("- from `%s` fn `%s` (%s) ref=%s ins=`%s`\n" % (
            rr[0], rr[1], rr[2], rr[4], rr[3].replace("`", "'")))
    f.write("\n")

f.write("## Function Call Graph (for functions seen in xrefs)\n\n")
for fentry in sorted(call_graph.keys()):
    fname, callers, callees = call_graph[fentry]
    f.write("### %s @ %s\n\n" % (fname, fentry))
    f.write("Callers:\n")
    if callers:
        for c in callers[:40]:
            f.write("- %s\n" % c)
    else:
        f.write("- (none)\n")
    f.write("Callees:\n")
    if callees:
        for c in callees[:80]:
            f.write("- %s\n" % c)
    else:
        f.write("- (none)\n")
    f.write("\n")

f.close()

print("Wrote: %s" % out_path)

File: docs/archive/low_trust/old_scripts/scripts/re/GhidraExtractFwFlow.java
// Extract firmware/DMA related string xrefs and surrounding functions for mtkwecx.sys
//@author codex
//@category MT7927

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.Data;
import ghidra.program.model.listing.DataIterator;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionIterator;
import ghidra.program.model.listing.Instruction;
import ghidra.program.model.listing.InstructionIterator;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.mem.MemoryAccessException;
import ghidra.program.model.symbol.Reference;
import ghidra.program.model.symbol.RefType;

public class GhidraExtractFwFlow extends GhidraScript {

    private static class RefRow {
        String from;
        String fnName;
        String fnEntry;
        String ins;
        String refType;
    }

    private static class MatchRow {
        String strAddr;
        String strVal;
        List<RefRow> refs = new ArrayList<>();
    }

    private static class FnGraph {
        String name;
        Set<String> callers = new HashSet<>();
        Set<String> callees = new HashSet<>();
    }

    @Override
    protected void run() throws Exception {
        String outPath = "/tmp/mtkwecx_fw_flow_auto.md";
        if (getScriptArgs().length > 0) {
            outPath = getScriptArgs()[0];
        }

        String[] keywords = new String[] {
            "MT6639PreFirmwareDownloadInit",
            "MT6639WpdmaConfig",
            "MT6639InitTxRxRing",
            "MT6639ConfigIntMask",
            "HalDownloadPatchFirmware",
            "AsicConnac3xLoadRomPatch",
            "AsicConnac3xLoadFirmware",
            "AsicConnac3xWpdmaInitRing",
            "AsicConnac3xWfdmaWaitIdle",
            "P0T15:AP CMD",
            "P0T16:FWDL",
            "P0R0:FWDL",
            "HOST_INT_STA",
            "MCU_CMD",
            "WPDMA_GLO_CFG",
            "WFDMA"
        };

        Listing listing = currentProgram.getListing();
        List<MatchRow> matches = new ArrayList<>();
        Set<String> seenStrAddr = new HashSet<>();

        DataIterator dit = listing.getDefinedData(true);
        while (dit.hasNext()) {
            Data d = dit.next();
            if (monitor.isCancelled()) {
                return;
            }
            Object v = d.getValue();
            if (v == null) {
                continue;
            }

            String s = String.valueOf(v);
            if (s.length() < 4) {
                continue;
            }

            boolean hit = false;
            for (String kw : keywords) {
                if (s.contains(kw)) {
                    hit = true;
                    break;
                }
            }
            if (!hit) {
                continue;
            }

            String saddr = d.getAddress().toString();
            if (seenStrAddr.contains(saddr)) {
                continue;
            }
            seenStrAddr.add(saddr);

            MatchRow mr = new MatchRow();
            mr.strAddr = saddr;
            mr.strVal = s;

            Reference[] refs = getReferencesTo(d.getAddress());
            for (Reference r : refs) {
                RefRow rr = new RefRow();
                Address from = r.getFromAddress();
                rr.from = from.toString();

                Function fn = getFunctionContaining(from);
                if (fn != null) {
                    rr.fnName = fn.getName();
                    rr.fnEntry = fn.getEntryPoint().toString();
                }
                else {
                    rr.fnName = "<no_function>";
                    rr.fnEntry = "";
                }

                Instruction ins = listing.getInstructionAt(from);
                rr.ins = ins != null ? ins.toString() : "";
                rr.refType = r.getReferenceType().toString();

                mr.refs.add(rr);
            }
            matches.add(mr);
        }

        Collections.sort(matches, Comparator.comparing(a -> a.strAddr));

        Map<String, FnGraph> fnMap = new HashMap<>();
        for (MatchRow mr : matches) {
            for (RefRow rr : mr.refs) {
                if (rr.fnEntry.isEmpty()) continue;
                FnGraph g = fnMap.get(rr.fnEntry);
                if (g == null) {
                    g = new FnGraph();
                    g.name = rr.fnName;
                    fnMap.put(rr.fnEntry, g);
                }
            }
        }

        for (Map.Entry<String, FnGraph> e : fnMap.entrySet()) {
            Address fEntry = toAddr(e.getKey());
            Function fn = getFunctionAt(fEntry);
            if (fn == null) continue;

            InstructionIterator iit = listing.getInstructions(fn.getBody(), true);
            while (iit.hasNext()) {
                Instruction ins = iit.next();
                if (ins.getFlowType().isCall()) {
                    Address[] flows = ins.getFlows();
                    if (flows != null) {
                        for (Address fa : flows) {
                            Function callee = getFunctionContaining(fa);
                            if (callee != null) {
                                e.getValue().callees.add(callee.getName() + "@" + callee.getEntryPoint());
                            }
                            else {
                                e.getValue().callees.add(fa.toString());
                            }
                        }
                    }
                }
            }

            Reference[] toRefs = getReferencesTo(fn.getEntryPoint());
            for (Reference r : toRefs) {
                if (!r.getReferenceType().isCall()) continue;
                Function caller = getFunctionContaining(r.getFromAddress());
                if (caller != null) {
                    e.getValue().callers.add(caller.getName() + "@" + caller.getEntryPoint());
                }
                else {
                    e.getValue().callers.add(r.getFromAddress().toString());
                }
            }
        }

        writeMarkdown(outPath, matches, fnMap);
        println("Wrote: " + outPath);
    }

    private void writeMarkdown(String outPath, List<MatchRow> matches, Map<String, FnGraph> fnMap)
            throws IOException {
        File out = new File(outPath);
        File parent = out.getParentFile();
        if (parent != null) parent.mkdirs();

        try (BufferedWriter bw = new BufferedWriter(new FileWriter(out))) {
            bw.write("# mtkwecx.sys firmware/DMA reverse notes (auto extract)\n\n");
            bw.write("Program: " + currentProgram.getName() + "\n\n");

            bw.write("## String Xrefs\n\n");
            for (MatchRow mr : matches) {
                bw.write("### " + mr.strAddr + " `" + esc(mr.strVal) + "`\n\n");
                if (mr.refs.isEmpty()) {
                    bw.write("- (no xrefs found)\n\n");
                    continue;
                }
                for (RefRow rr : mr.refs) {
                    bw.write("- from `" + rr.from + "` fn `" + esc(rr.fnName) + "` (" + rr.fnEntry + ") ref="
                            + esc(rr.refType) + " ins=`" + esc(rr.ins) + "`\n");
                }
                bw.write("\n");
            }

            bw.write("## Function Call Graph (functions seen in xrefs)\n\n");
            List<String> keys = new ArrayList<>(fnMap.keySet());
            Collections.sort(keys);
            for (String entry : keys) {
                FnGraph g = fnMap.get(entry);
                bw.write("### " + esc(g.name) + " @ " + entry + "\n\n");
                bw.write("Callers:\n");
                if (g.callers.isEmpty()) {
                    bw.write("- (none)\n");
                } else {
                    List<String> callers = new ArrayList<>(g.callers);
                    Collections.sort(callers);
                    for (String c : callers) bw.write("- " + esc(c) + "\n");
                }
                bw.write("Callees:\n");
                if (g.callees.isEmpty()) {
                    bw.write("- (none)\n");
                } else {
                    List<String> callees = new ArrayList<>(g.callees);
                    Collections.sort(callees);
                    for (String c : callees) bw.write("- " + esc(c) + "\n");
                }
                bw.write("\n");
            }
        }
    }

    private String esc(String s) {
        if (s == null) return "";
        return s.replace("`", "'").replace("\n", "\\n").replace("\r", "");
    }
}
